### ISSUE TITLE
Upgrading makara: client side

### ISSUE THREAD

**Issue by Shino-** [krakenjs/makara#78]
In the ["upgrading makara" guide](http://krakenjs.com/2015/07/09/upgrading-makara), at the end it's written:

> Adjust your client accordingly, and use [dust-usecontent-helper](https://github.com/krakenjs/dust-usecontent-helper) and [dust-message-helper](https://github.com/krakenjs/dust-message-helper) or [dust-intl](http://formatjs.io/dust) to load the content for your templates.

But the README of these doesn't tell us how to use them in the browser so I'm a bit lost. Either these READMEs should be updated with client side examples or the upgrading makara guide should be a little bit more detailed (ideally both).

In the meantime, I'll continue investigating so maybe this issue will be transformed in a pull request in the helpers repo.

**Comment by grawk** [comment]
Yes that's a very good point. I do have a project which is currently only available within the PayPal corporate github which illustrates this solution. I will work to make it available out in the open this week.

**Comment by grawk** [comment]
Here is my sample app, with detailed README on using makara in the browser:
https://github.com/grawk/try18n/blob/master/README.md

It needs to be gone through. It's basically a first draft. Let me know if it helps, and/or if there are any inaccuracies (if you have time to look through it).

Hopefully, you can clone that repo, install, and it should Just Work.

**Comment by saccodd** [comment]
I'm trying to implement I18N on client side as well, but using browserify.
Do you have any tutorial or example about? Actually I'm stuck and I cannot even consume the language pack built by grunt and browserify. Following code returns an empty object for my test language pack, when actually it is not empty: 

``` javascript
//language pack loaded with browserify 
var langPack=require('../../.build/en-US/_languagepack.js');

dust.onLoad = function(templateName, callback) {
    console.log("loading template"+ templateName);
    $.get('/templates/' + templateName + '.js', function(data) {
        var res=dust.loadSource(data);
        callback(null,res);
    });
};

var loader = function(ctx, bundle, callback){
    console.log('loader ctx '+JSON.stringify(ctx)+' bundle'+bundle);
    console.log(JSON.stringify(langPack,0,4));  // is an empty object
    callback(null,langPack[getLang()][bundle]);
};

require('dust-makara-helpers').registerWith(dust, {
    enableMetadata: true,
    autoloadTemplateContent: false,
    loader:loader
});

function getLang() {
    console.log(document.documentElement.getAttribute('lang'))
    return document.documentElement.getAttribute('lang');
}
```

I'm missing something for sure. My kraken.js back-end is configured with makara-browserify and works fine. Maybe I should include the distribution for the client but then I wouldn't know how to use it in my code shown above.

**Comment by Shino-** [comment]
@saccodd : Be careful with `require` and `..` paths, usually I use `__dirname` otherwise the `../../` path depends on the place where you launch your command.

**Comment by aredridel** [comment]
Not actually true. require is always relative to `__dirname` :)

**Comment by saccodd** [comment]
It wasn't a path issue. The language pack can't be required within my app.js, otherwise grunt-browserify will try to build it in the bundle (and actually it doesn't make sense as language packs should be decoupled from the app bundle). Also, the language pack is built by makara-browserify by using the "exposed" flavor, instead of wrapping it in a function. 
I'm a newbie with Browserify but I came out with the following solution. 
The language pack is dynamically loaded in my template, according to user locale:

``` html
<script src="{context.links.resourceBaseUrl|s}/{makara.languagePackPath|s}" ></script>
<script>
     var langPack = require('_languagepack');
</script>
<script type="text/javascript" src="js/app-bundle.js" ></script>
```

Now langPack variable can be consumed in my app-bundle.js. It's slightly changed and I attach it for sake of completeness.

``` javascript
console.log(langPack); // GOT IT!!

dust.onLoad = function(templateName, callback) {
    console.log("loading template"+ templateName);
    $.get('/templates/' + templateName + '.js', function(data) {
        var res=dust.loadSource(data);
        callback(null,res);
    });
};

var loader = function(ctx, bundle, callback){
    console.log('loader ctx '+JSON.stringify(ctx)+' bundle'+bundle);
    console.log(JSON.stringify(langPack,0,4));
    callback(null,langPack[getLang()][bundle]);
};

require('dust-makara-helpers').registerWith(dust, {
    enableMetadata: true,
    autoloadTemplateContent: false,
    loader:loader
});

function getLang() {
    console.log(document.documentElement.getAttribute('lang'))
    return document.documentElement.getAttribute('lang');
}
```

It works. However, can you tell me if this trick has any drawback in your opinion? if language pack size is large, a potential performance issue may arise. The language pack is always loaded by the browser, even if no client side rendering will be performed. However, it's something that I can tolerate for my application because language packs will be rather small.

**Comment by grawk** [comment]
@saccodd I'm trying to add a branch onto my `try18n` application to exemplify browserify usage. If you've still got it handy, can you share how you've configured dustjs to load in the browserify in your setup?

**Comment by grawk** [comment]
I've added a `browserify` branch. I've used your `app-bundle.js` file as a basis, as well as your approach to loading the langpack:
https://github.com/grawk/try18n/tree/browserify

I'm not able to discern any issues with including the langpack as you've done. One possible efficiency could be to modify the `makara` middleware to provide the current langpack JSON as a string. Then it could be included directly in the server-rendered template and remove the need for an additional HTTP request.
